### PR TITLE
Add post-dominators

### PR DESCRIPTION
**Pull request by kevinbackhouse**
Adding support for post-dominators to the Dominance library.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/controlflow/Dominance.qll [hunk 1]
```diff
@@ -57,6 +85,14 @@ predicate dominates(ControlFlowNode dominator, ControlFlowNode node) {
   strictlyDominates(dominator, node) or dominator = node
 }
 
+/**
+ * Holds if `postdominator` is a post-dominator of `node` in the control-flow graph. This
+ * is reflexive.
+ */
+predicate postdominates(ControlFlowNode postdominator, ControlFlowNode node) {
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/controlflow/Dominance.qll]: This should be `postDominates` (upper-case D).

File: cpp/ql/src/semmle/code/cpp/controlflow/Dominance.qll [hunk 2]
```diff
@@ -67,6 +103,25 @@ predicate dominates(ControlFlowNode dominator, ControlFlowNode node) {
...
 
+/**
+ * Holds if `pred` is a predecessor of `succ` in the control-flow graph of
+ * basic blocks.
+ */
+private predicate bb_predecessor(BasicBlock succ, BasicBlock pred) {
+  bb_successor(pred, succ)
+}
+
+/** Holds if `exit` an `ExitBasicBlock`. */
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/controlflow/Dominance.qll]: Missing "is"?

File: cpp/ql/src/semmle/code/cpp/controlflow/Dominance.qll [hunk 3]
```diff
@@ -67,6 +103,25 @@ predicate dominates(ControlFlowNode dominator, ControlFlowNode node) {
...
+private predicate bb_exit(ExitBasicBlock exit) {
+  any()
+}
+
+/**
+ * Holds if `postdominator` is an immediate post-dominator of `node` in the control-flow
+ * graph of basic blocks.
+ */
+predicate bbIPostDominates(BasicBlock dom, BasicBlock node) = idominance(bb_exit/1, bb_predecessor/2)(_, dom, node)
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/controlflow/Dominance.qll]: `dom` should be `postdominator` to match the QLDoc and similar predicates.

**Comment by aschackmull** [comment]
If there are exception edges reaching the exit-node, then what's usually needed when reasoning about post-dominance is post-dominance under the assumption of normal termination of the enclosing function.  This can be achieved by adding another exit-node to the cfg that collects normal return edges before joining exception edges at the final exit-node and calculating post-dominance based on that instead.  This might not be very relevant for C++, but I've been thinking about adding this additional notion of post-dominance to the java libs, since regular post-dominance doesn't work well there.

**Comment by kevinbackhouse** [comment]
@aschackmull: my main use-case for this is to check that something is cleaned up before the function exits. For example, checking that a mutex is always unlocked before the function exits. Or checking that a file descriptor is always closed before the function exits. So I think I would want to include things like exception edges. But perhaps not things like calls to `exit`/`abort` because it's not such a big deal if the program is going to terminate anyway.

**Pull request merged**